### PR TITLE
LVIIICNET-195 sending an email when registering

### PR DESCRIPTION
**Pull request by aaandrianaaa**

### CODE REVIEW

File: Common/Configurations/TokenManagement.cs [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+
+        [JsonProperty("audience")]
+        public string Audience { get; set; }
+
+        [JsonProperty("accessExpiration")]
+        public int AccessExpiration { get; set; }
+
+        [JsonProperty("refreshExpiration")]
+        public int RefreshExpiration { get; set; }
+    }
```
Comment (mod1fy) [inline Common/Configurations/TokenManagement.cs]: Do we need token settings in common?

File: TheraLang.Web/ClientApp/src/app/core/auth/auth.service.ts [hunk 2]
```diff
@@ -91,4 +91,11 @@ export class AuthService {
       responseType: "text"
     });
   }
+
+  confirmUser(confirmationNumber: number, email: string) {
+    return this.http.post(this.baseUrl + "/confirmation", {
+        confirmation_number: confirmationNumber,
+        email,
+      });
```
Comment (mod1fy) [inline TheraLang.Web/ClientApp/src/app/core/auth/auth.service.ts]:
```suggestion
  confirmUser(confirmationNumber: number, email: string) {
    return this.http.post(`${this.baseUrl}/confirmation`, {
      confirmation_number: confirmationNumber,
      email,
    });
```

File: TheraLang.BLL/Services/UserManagementService.cs [hunk 3]
```diff
@@ -86,5 +94,56 @@ public async Task AddUser(UserAllDto NewUser)
                 throw new Exception("Error when adding user ", ex);
             }
         }
+
+        public async Task SendEmail(int ConfirmNum, string UserEmail)
+        {
+            try
+            {
+                string body = string.Empty;
```
Comment (mod1fy) [inline TheraLang.BLL/Services/UserManagementService.cs]: var?

File: TheraLang.BLL/Services/UserManagementService.cs [hunk 4]
```diff
@@ -86,5 +94,56 @@ public async Task AddUser(UserAllDto NewUser)
...
+                using (StreamReader reader = new StreamReader(_env.ContentRootPath + "/Templates/welcome.html"))
```
Comment (mod1fy) [inline TheraLang.BLL/Services/UserManagementService.cs]:
```suggestion
                using (StreamReader reader =
                    new StreamReader(Path.Combine(_env.ContentRootPath, "Templates", "welcome.html")))
```

File: TheraLang.BLL/Services/UserManagementService.cs [hunk 5]
```diff
@@ -86,5 +94,56 @@ public async Task AddUser(UserAllDto NewUser)
...
+                    body = reader.ReadToEnd();
+                }
+
+                var fromAddress = new MailAddress(_emailSettings.Email , "UTMM");
+                var toAddress = new MailAddress(UserEmail, "To User");
+                string fromPassword = _emailSettings.Password;
+                string subject = "Confirm your email";
+
+                body = body.Replace("{EMAIL}", UserEmail);
+                body = body.Replace("{NUMBER}", ConfirmNum.ToString());
```
Comment (mod1fy) [inline TheraLang.BLL/Services/UserManagementService.cs]: Maybe string.format()?

File: TheraLang.BLL/Services/UserManagementService.cs [hunk 6]
```diff
@@ -86,5 +94,56 @@ public async Task AddUser(UserAllDto NewUser)
...
         }
+
+        public async Task SendEmail(int ConfirmNum, string UserEmail)
+        {
+            try
+            {
+                string body = string.Empty;
+                using (StreamReader reader = new StreamReader(_env.ContentRootPath + "/Templates/welcome.html"))
+                {
+                    body = reader.ReadToEnd();
```
Comment (mod1fy) [inline TheraLang.BLL/Services/UserManagementService.cs]:
```suggestion
                    body = await reader.ReadToEndAsync();
```

File: TheraLang.BLL/Services/UserManagementService.cs [hunk 7]
```diff
@@ -86,5 +94,56 @@ public async Task AddUser(UserAllDto NewUser)
...
+                    Credentials = new NetworkCredential(fromAddress.Address, fromPassword)
+                };
+                using (var message = new MailMessage(fromAddress, toAddress)
+                {
+                    Subject = subject,
+                    IsBodyHtml = true,
+                    Body = body,
+                })
+                {
+                    smtp.Send(message);
```
Comment (mod1fy) [inline TheraLang.BLL/Services/UserManagementService.cs]:
```suggestion
                    await smtp.SendMailAsync(message);
```

File: TheraLang.BLL/Services/UserManagementService.cs [hunk 8]
```diff
@@ -86,5 +94,56 @@ public async Task AddUser(UserAllDto NewUser)
...
+                }
+            }
+            catch (Exception ex)
+            {
+                throw new Exception("Error when sending email ", ex);
+            }
```
Comment (mod1fy) [inline TheraLang.BLL/Services/UserManagementService.cs]: Don't catch/throw Exception. Create custom one derived from ApiException

File: TheraLang.DAL/Entities/User.cs [hunk 9]
```diff
@@ -10,6 +10,8 @@ public class User
         public string PasswordHash { get; set; }
         public Guid RoleId { get; set; }
         public Role Role { get; set; }
+        public int ConfirmationNumber { get; set; }
+        public bool IsConfirmByEmail { get; set; } = false;
```
Comment (mod1fy) [inline TheraLang.DAL/Entities/User.cs]: Maybe role?

File: Common/Configurations/EmailSettings.cs [hunk 10]
```diff
@@ -0,0 +1,14 @@
...
+{
+    [JsonObject("emailSettings")]
+    public class EmailSettings
+    {
+        [JsonProperty("email")]
+        public string Email { get; set; }
+
+        [JsonProperty("password")]
+        public string Password { get; set; }
+    }
```
[the remark this answers is not in the document]
Comment (aaandrianaaa) [reply Common/Configurations/EmailSettings.cs]: done

File: TheraLang.BLL/Services/UserManagementService.cs [hunk 11]
```diff
@@ -70,12 +79,11 @@ public async Task AddUser(UserAllDto NewUser)
                     }
 
                     var user = mapper.Map<UserAllDto, User>(NewUser);
-
                     user.RoleId = (await _unitOfWork.Repository<Role>().Get(r => r.Name == "Guest")).Id;
                     user.PasswordHash = PasswordHasher.HashPassword(NewUser.Password);
                     user.Details = userDetails;
 
-                    _unitOfWork.Repository<UserDetails>().Add(userDetails);
+                    //_unitOfWork.Repository<UserDetails>().Add(userDetails);
```
[the remark this answers is not in the document]
Comment (aaandrianaaa) [reply TheraLang.BLL/Services/UserManagementService.cs]: done

File: TheraLang.BLL/Services/UserManagementService.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (aaandrianaaa) [reply TheraLang.BLL/Services/UserManagementService.cs]: done

File: TheraLang.BLL/Services/UserManagementService.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (aaandrianaaa) [reply TheraLang.BLL/Services/UserManagementService.cs]: done

File: TheraLang.BLL/DataTransferObjects/UserDtos/ConfirmUserDto.cs [hunk 12]
```diff
@@ -0,0 +1,12 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
```
[the remark this answers is not in the document]
Comment (aaandrianaaa) [reply TheraLang.BLL/DataTransferObjects/UserDtos/ConfirmUserDto.cs]: done

File: Common/Configurations/TokenManagement.cs (hunk 1 above)
Comment (aaandrianaaa) [reply Common/Configurations/TokenManagement.cs]: I think we shoud save all our configuration classes in one plase in progect, separately from Dtos

File: TheraLang.BLL/Services/UserManagementService.cs (hunk 7 above)
[the remark this answers is not in the document]
Comment (aaandrianaaa) [reply TheraLang.BLL/Services/UserManagementService.cs]: done

File: TheraLang.DAL/Entities/User.cs (hunk 9 above)
Comment (aaandrianaaa) [reply TheraLang.DAL/Entities/User.cs]: okay, I'll create another pull request with creating new roles

File: TheraLang.Web/ClientApp/src/app/core/auth/auth.service.ts (hunk 2 above)
Comment (aaandrianaaa) [reply TheraLang.Web/ClientApp/src/app/core/auth/auth.service.ts]: done

File: TheraLang.BLL/Services/UserManagementService.cs (hunk 8 above)
[the remark this answers is not in the document]
Comment (aaandrianaaa) [reply TheraLang.BLL/Services/UserManagementService.cs]: done

File: TheraLang.DAL/Entities/User.cs (hunk 9 above)
Comment (aaandrianaaa) [reply TheraLang.DAL/Entities/User.cs]: done

File: TheraLang.DAL/Entities/UserConfirmation.cs [hunk 13]
```diff
@@ -0,0 +1,14 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
```
Comment (oleksiistetsiouk) [inline TheraLang.DAL/Entities/UserConfirmation.cs]: usings

File: TheraLang.DAL/Entities/UserDetails.cs [hunk 14]
```diff
@@ -17,7 +17,8 @@ public UserDetails()
         public string ShortInformation { get; set; }
         public string ImageURl { get; set; }
         public User User { get; set; }
-        public int Age { get; set; }
+        public DateTime BirthdayDate { get; set; }
```
Comment (oleksiistetsiouk) [inline TheraLang.DAL/Entities/UserDetails.cs]: BirthDate maybe? (99)

File: TheraLang.Web/ClientApp/src/assets/i18n/ua.json [hunk 15]
```diff
@@ -111,6 +111,7 @@
     },
     "account": {
       "signUp": "Реєстрація",
+      "forgotPassword": "Заубли пароль?",
```
Comment (oleksiistetsiouk) [inline TheraLang.Web/ClientApp/src/assets/i18n/ua.json]: *Забули

File: TheraLang.Web/ClientApp/src/assets/i18n/ua.json [hunk 16]
```diff
@@ -131,7 +132,19 @@
...
+      "redirection-to-login": "Перейти на логін",
+      "error": "Помилка",
+      "password-forgot": {
+        "title": "Введіть ваш емейл",
+        "success": "Ми відправили Вам емейл з наступними кроками!",
+        "error": "Користувача з даним емейлом не знайдено!"
+      },
+      "password-reset": {
+        "title": "Введіть новий пароль",
+        "error": "Неможливо обновити пароль!"
```
Comment (oleksiistetsiouk) [inline TheraLang.Web/ClientApp/src/assets/i18n/ua.json]: *оновити

File: TheraLang.Web/Controllers/AccountController.cs [hunk 17]
```diff
@@ -29,7 +33,7 @@ public async Task<IActionResult> SignIn([FromBody] LoginModel login)
             var loginDto = mapper.Map<LoginModel, LoginModelDto>(login);
             var user = await _userManagement.GetUser(loginDto.Email, loginDto.Password);
             var token = await _authService.Authenticate(user);
-            if (token == "")
+            if (token == "" || token == null)
```
Comment (oleksiistetsiouk) [inline TheraLang.Web/Controllers/AccountController.cs]: use (String.IsNotNullOrEmpy(token))

File: TheraLang.Web/ViewModels/UsersViewModels/ConfirmUserViewModel.cs [hunk 18]
```diff
@@ -0,0 +1,16 @@
+﻿using Newtonsoft.Json;
+using System;
+using System.Collections.Generic;
```
Comment (oleksiistetsiouk) [inline TheraLang.Web/ViewModels/UsersViewModels/ConfirmUserViewModel.cs]: usings

File: TheraLang.Web/ViewModels/UsersViewModels/UserAllViewModel.cs [hunk 19]
```diff
@@ -21,6 +22,6 @@ public class UserAllViewModel
         public string LastName { get; set; }
 
         public IFormFile Image { get; set; }
-        public int Age { get; set; }
+        public DateTime BirthdayDate { get; set; }
```
Comment (oleksiistetsiouk) [inline TheraLang.Web/ViewModels/UsersViewModels/UserAllViewModel.cs]: BirthDate maybe? (90210)

File: TheraLang.Web/ViewModels/UsersViewModels/UserDetailsViewModel.cs [hunk 20]
```diff
@@ -9,5 +10,6 @@ public class UserDetailsViewModel
         public IFormFile Image { get; set; }
         public string PhoneNumber { get; set; }
         public string ShortInformation { get; set; }
+        public DateTime BirthdayDate { get; set; }
```
Comment (oleksiistetsiouk) [inline TheraLang.Web/ViewModels/UsersViewModels/UserDetailsViewModel.cs]: BirthDate maybe? (999999)

File: TheraLang.Web/ViewModels/UsersViewModels/UsersViewModel.cs [hunk 21]
```diff
@@ -1,10 +1,13 @@
...
+
+namespace TheraLang.Web.ViewModels.UsersViewModels
 {
     public class UsersViewModel
     {
         public string FirstName { get; set; }
         public string LastName { get; set; }
         public string ShortInformation { get; set; }
         public string ImageURl { get; set; }
+        public DateTime BirthdayDate { get; set; }
```
Comment (oleksiistetsiouk) [inline TheraLang.Web/ViewModels/UsersViewModels/UsersViewModel.cs]: BirthDate maybe? (last time)

File: TheraLang.BLL/Services/UserManagementService.cs [hunk 22]
```diff
@@ -95,55 +103,24 @@ public async Task AddUser(UserAllDto NewUser)
...
+                int random = rand.Next(10000000, 100000000);
+                conf.Number = random;
+                conf.ConfDateTime = DateTime.Now;
+                await _unitOfWork.SaveChangesAsync();
+                await _confirmation.SendEmail(random.ToString(), Email, "password.html");
             }
             catch (Exception ex)
             {
                 throw new Exception("Error when sending email ", ex);
             }
```
Comment (mod1fy) [inline TheraLang.BLL/Services/UserManagementService.cs]: XXX

File: TheraLang.DAL/IttmmDbContext.cs [hunk 23]
```diff
@@ -29,6 +30,9 @@ public IttmmDbContext(DbContextOptions options) : base(options)
         public virtual DbSet<PageRoute> PageRoutes { get; set; }
         public virtual DbSet<UserDetails> UsersDetails { get; set; }
         public virtual DbSet<MemberFee> MemberFees { get; set; }
+        public virtual DbSet<PaymentHistory> PaymentHistory { get; set; }
+        public virtual DbSet<NewsComment> NewsComments { get; set; }
+        public virtual DbSet<NewsLike> NewsLikes { get; set; }
```
Comment (mod1fy) [inline TheraLang.DAL/IttmmDbContext.cs]: Dbset for UserConfirmation?

File: TheraLang.DAL/IttmmDbContext.cs [hunk 24]
```diff
@@ -42,12 +46,15 @@ protected override void OnModelCreating(ModelBuilder modelBuilder)
...
+            modelBuilder.ApplyConfiguration(new NewsLikesConfiguration());
             modelBuilder.ApplyConfiguration(new UploadedNewsContentImageConfiguration());
             modelBuilder.ApplyConfiguration(new UserConfiguration());
             modelBuilder.ApplyConfiguration(new UserDetailsConfiguration());
             modelBuilder.ApplyConfiguration(new PageConfiguration());
             modelBuilder.ApplyConfiguration(new PageRouteConfiguration());
             modelBuilder.ApplyConfiguration(new MemberFeeConfiguration());
+            modelBuilder.ApplyConfiguration(new PaymentHistoryConfiguration()); 
+            modelBuilder.ApplyConfiguration(new NewsCommentConfiguration());
         }
```
Comment (mod1fy) [inline TheraLang.DAL/IttmmDbContext.cs]: and configuration?

File: DataSeeding/DatabaseSeed.cs [hunk 25]
```diff
@@ -15,7 +15,7 @@ public static IWebHost Seed(this IWebHost host)
                     var dataseed = new DbInitializer();
                     dataseed.Seed(scope);
                 }
-                catch (Exception ex)
+                catch (Exception)
                 {
                     //logger.LogError(ex, "An error occurred seeding the DB.");
```
[the remark this answers is not in the document]
Comment (aaandrianaaa) [reply DataSeeding/DatabaseSeed.cs]: done

File: TheraLang.BLL/DataTransferObjects/UserDtos/ConfirmPasswordChangingDto.cs [hunk 26]
```diff
@@ -0,0 +1,14 @@
+﻿using System;
+using System.Collections.Generic;
```
[the remark this answers is not in the document]
Comment (aaandrianaaa) [reply TheraLang.BLL/DataTransferObjects/UserDtos/ConfirmPasswordChangingDto.cs]: done

File: TheraLang.BLL/DataTransferObjects/UserDtos/UserAllDto.cs [hunk 27]
```diff
@@ -11,5 +12,8 @@ public class UserAllDto
         public string Password { get; set; }
         public string ShortInformation { get; set; }
         public string ImageURl { get; set; }
+        public int ConfirmationNumber { get; set; }
+        public bool IsConfirmByEmail { get; set; }
+        public DateTime BirthdayDate { get; set; }
```
[the remark this answers is not in the document]
Comment (aaandrianaaa) [reply TheraLang.BLL/DataTransferObjects/UserDtos/UserAllDto.cs]: done

File: TheraLang.BLL/DataTransferObjects/UserDtos/UserDetailsDto.cs [hunk 28]
```diff
@@ -10,5 +11,6 @@ public class UserDetailsDto
         public string PhoneNumber { get; set; }
         public string ShortInformation { get; set; }
         public string ImageURl { get; set; }
+        public DateTime BirthdayDate { get; set; }
```
[the remark this answers is not in the document]
Comment (aaandrianaaa) [reply TheraLang.BLL/DataTransferObjects/UserDtos/UserDetailsDto.cs]: done

File: TheraLang.BLL/DataTransferObjects/UserDtos/UsersListDto.cs [hunk 29]
```diff
@@ -0,0 +1,12 @@
+﻿using System;
+using System.Collections.Generic;
```
[the remark this answers is not in the document]
Comment (aaandrianaaa) [reply TheraLang.BLL/DataTransferObjects/UserDtos/UsersListDto.cs]: done

File: TheraLang.BLL/Interfaces/IUserManagementService.cs [hunk 30]
```diff
@@ -12,5 +12,6 @@ public interface IUserManagementService
         Task<User> GetUserById(Guid id);
 
         Task AddUser(UserAllDto NewUser);
+        Task PasswordConfirmationRequest(string Email);
```
[the remark this answers is not in the document]
Comment (aaandrianaaa) [reply TheraLang.BLL/Interfaces/IUserManagementService.cs]: done

File: TheraLang.BLL/Services/ConfirmationService.cs [hunk 31]
```diff
@@ -0,0 +1,99 @@
+﻿using Microsoft.AspNetCore.Hosting;
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Text;
+using System.Threading.Tasks;
+using System.Net;
+using System.Net.Mail;
+using TheraLang.DAL.UnitOfWork;
```
[the remark this answers is not in the document]
Comment (aaandrianaaa) [reply TheraLang.BLL/Services/ConfirmationService.cs]: done

File: TheraLang.DAL/Configuration/UserConfirmationConfiguration.cs [hunk 32]
```diff
@@ -0,0 +1,18 @@
...
+using TheraLang.DAL.Entities;
+
+namespace TheraLang.DAL.Configuration
+{
+        internal class UserConfirmationConfiguration : IEntityTypeConfiguration<UserConfirmation>
+        {
+            public void Configure(EntityTypeBuilder<UserConfirmation> builder)
+            {
+                builder.ToTable("UsersConfirmations");
+
```
[the remark this answers is not in the document]
Comment (aaandrianaaa) [reply TheraLang.DAL/Configuration/UserConfirmationConfiguration.cs]: done

File: TheraLang.DAL/Configuration/UserConfirmationConfiguration.cs [hunk 33]
```diff
@@ -0,0 +1,18 @@
+﻿using Microsoft.EntityFrameworkCore;
+using Microsoft.EntityFrameworkCore.Metadata.Builders;
+using System;
+using System.Collections.Generic;
+using System.Text;
```
[the remark this answers is not in the document]
Comment (aaandrianaaa) [reply TheraLang.DAL/Configuration/UserConfirmationConfiguration.cs]: done

File: TheraLang.DAL/Entities/UserConfirmation.cs (hunk 13 above)
Comment (aaandrianaaa) [reply TheraLang.DAL/Entities/UserConfirmation.cs]: done

File: TheraLang.DAL/Entities/UserDetails.cs (hunk 14 above)
Comment (aaandrianaaa) [reply TheraLang.DAL/Entities/UserDetails.cs]: done

File: TheraLang.Web/ClientApp/src/assets/i18n/ua.json (hunk 15 above)
Comment (aaandrianaaa) [reply TheraLang.Web/ClientApp/src/assets/i18n/ua.json]: done

File: TheraLang.Web/ClientApp/src/assets/i18n/ua.json (hunk 16 above)
Comment (aaandrianaaa) [reply TheraLang.Web/ClientApp/src/assets/i18n/ua.json]: done

File: TheraLang.Web/ViewModels/UsersViewModels/ConfirmUserViewModel.cs (hunk 18 above)
Comment (aaandrianaaa) [reply TheraLang.Web/ViewModels/UsersViewModels/ConfirmUserViewModel.cs]: done

File: TheraLang.Web/ViewModels/UsersViewModels/UserAllViewModel.cs (hunk 19 above)
Comment (aaandrianaaa) [reply TheraLang.Web/ViewModels/UsersViewModels/UserAllViewModel.cs]: done

File: TheraLang.Web/ViewModels/UsersViewModels/UserDetailsViewModel.cs (hunk 20 above)
Comment (aaandrianaaa) [reply TheraLang.Web/ViewModels/UsersViewModels/UserDetailsViewModel.cs]: done

File: TheraLang.Web/ViewModels/UsersViewModels/UsersViewModel.cs (hunk 21 above)
Comment (aaandrianaaa) [reply TheraLang.Web/ViewModels/UsersViewModels/UsersViewModel.cs]: done

File: TheraLang.BLL/Services/UserManagementService.cs (hunk 22 above)
Comment (aaandrianaaa) [reply TheraLang.BLL/Services/UserManagementService.cs]: done

File: TheraLang.DAL/IttmmDbContext.cs (hunk 23 above)
Comment (aaandrianaaa) [reply TheraLang.DAL/IttmmDbContext.cs]: done

File: TheraLang.DAL/IttmmDbContext.cs (hunk 24 above)
Comment (aaandrianaaa) [reply TheraLang.DAL/IttmmDbContext.cs]: done

**Pull request merged**